flexmeasures/data/migrations/versions/f0ee99278f6f_added_flexmodel_to_generic_asset_model.py

```diff
@@ -0,0 +1,303 @@
+
+from flexmeasures.utils.unit_utils import is_power_unit, is_energy_unit, ur
+
+# revision identifiers, used by Alembic.
+revision = "f0ee99278f6f"
+down_revision = "cb8df44ebda5"
+branch_labels = None
+depends_on = None
+
+flex_model_fields = dict(
+    min_soc_in_mwh="soc-min",
+)
+
+
+def group_sensors_by_field(sensors, conn, generic_asset_table):
+    """
+    This function groups sensors by flexmodel fields (using the old field names) and checks for value mismatches.
+    """
+    field_specs = []
+
+    # construct the field specifications
+    for old_field_name, new_field_name in flex_model_fields.items():
+        field_spec = dict(
+            new_field_name=new_field_name,
+            old_field_name=old_field_name,
+            grouped=defaultdict(list),
+        )
+        field_specs.append(field_spec)
+
+    # iterate over the sensors and group them by field
+    # and check for value mismatches
+    for sensor in sensors:
+        # fetch the generic asset
+        sensor_generic_asset = conn.execute(
+            sa.select(
+                generic_asset_table.c.id,
+                generic_asset_table.c.attributes,
+            ).where(generic_asset_table.c.id == sensor.generic_asset_id)
+        ).fetchone()
+
+        if sensor_generic_asset is None:
+            raise Exception(
+                f"Generic asset not found for sensor {sensor.id} with asset_id {sensor.generic_asset_id}"
+            )
+
+        sensor_attrs = sensor.attributes or {}
+        asset_flex_model = sensor_generic_asset.attributes.get("flex-model", {})
+
+        for field_spec in field_specs:
+            old_field_name = field_spec["old_field_name"]
+
+            # check if old_field_name exist on both attributes on sensor and asset
+            sensor_val = sensor_attrs.get(old_field_name)
+            asset_val = asset_flex_model.get(old_field_name)
+
+            if sensor_val is not None and asset_val is not None:
+                if sensor_val != asset_val:
+                    raise Exception(
+                        f"Value mismatch for '{old_field_name}' in sensor {sensor.id}: sensor={sensor_val}, asset={asset_val}"
+                    )
+
+            # check if old_field_name exist on sensor attributes
+            if sensor_val is not None:
+                field_spec["grouped"][sensor_generic_asset].append(sensor)
+
+    return field_specs
+
+
+def validate_for_duplicate_keys(fields_specs):
+    """
+    This function checks for duplicate keys in the grouped sensors of a parent asset.
+    """
+    for field_spec in fields_specs:
+        grouped = field_spec["grouped"]
+        for asset_id, sensors in grouped.items():
+            if len(sensors) > 1:
+                raise Exception(
+                    f"Multiple sensors with '{field_spec['old_field_name']}' found for asset_id {asset_id}: {[s.id for s in sensors]}"
+                )
+
+
+def upgrade():
+    # ### commands auto generated by Alembic - please adjust! ###
```

Comment:
Please remove these auto generated comments (there are 4 in this file) to indicate to future devs that we actually adjusted the auto generated commands.